packages/jupyter-ai/jupyter_ai/chat_handlers/learn.py

```diff
@@ -153,19 +158,22 @@ async def learn_dir(
             default_splitter=RecursiveCharacterTextSplitter(**splitter_kwargs),
         )
 
-        delayed = split(path, all_files, splitter=splitter)
-        doc_chunks = await dask_client.compute(delayed)
-
-        em_provider_cls, em_provider_args = self.get_embedding_provider()
-        delayed = get_embeddings(doc_chunks, em_provider_cls, em_provider_args)
-        embedding_records = await dask_client.compute(delayed)
-        if self.index:
-            self.index.add_embeddings(*embedding_records)
-        else:
-            self.create(*embedding_records)
-
-        self._add_dir_to_metadata(path, chunk_size, chunk_overlap)
-        self.prev_em_id = em_provider_cls.id + ":" + em_provider_args["model_id"]
+        try:
+            delayed = split(path, all_files, splitter=splitter)
+            doc_chunks = await dask_client.compute(delayed)
+            em_provider_cls, em_provider_args = self.get_embedding_provider()
+            delayed = get_embeddings(doc_chunks, em_provider_cls, em_provider_args)
+            embedding_records = await dask_client.compute(delayed)
+            if self.index:
+                self.index.add_embeddings(*embedding_records)
+            else:
+                self.create(*embedding_records)
+
+            self._add_dir_to_metadata(path, chunk_size, chunk_overlap)
+            self.prev_em_id = em_provider_cls.id + ":" + em_provider_args["model_id"]
+        except Exception as e:
```

Comment:
`Exception` seems to very broad, should we limit this to `ModuleNotFoundError` which limits this to import errors? Also, once the error is caught, why don't we try replying in the same session without the user making another attempt before they see the response.

Reply:
After trying other error types such as `ModuleNotFoundError` it does not work because there are other exceptions related to embedding errors, etc. The only one that works is `Exception`. See `learn.py` also uses `Exception` on line 79 as well for embedding related issues. In the class `LearnChatHandler` the module `learn_dir` throws errors for other cases such as embeddings and not only for missing packages. 

Your second suggestion about replying in the same session is now implemented around the failure of `learn_dir` which also means that `Exception` is needed.